assembly/src/ast/module.rs

```diff
@@ -1,321 +1,718 @@
+    diagnostics::{Report, SourceFile},
+    sema::SemanticAnalysisError,
+    ByteReader, ByteWriter, Deserializable, DeserializationError, LibraryNamespace, LibraryPath,
+    Serializable, SourceSpan, Span, Spanned,
+};
+use alloc::{
+    boxed::Box,
+    string::{String, ToString},
+    sync::Arc,
+    vec::Vec,
 };
-use crate::utils::{collections::*, string::*};
+use core::fmt;
 
-use core::{fmt, str::from_utf8};
-use vm_core::utils::Serializable;
+use super::{Export, Import, LocalNameResolver, ProcedureIndex, ProcedureName, ResolvedProcedure};
 
-// MODULE AST
+// MODULE KIND
 // ================================================================================================
 
-/// An abstract syntax tree of a Miden module.
+/// Represents the kind of a [Module].
 ///
-/// A module AST consists of a list of procedure ASTs, a list of re-exported procedures, a list of
-/// imports, and module documentation. Local procedures could be internal or exported.
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub struct ModuleAst {
-    pub(super) local_procs: Vec<ProcedureAst>,
-    pub(super) reexported_procs: Vec<ProcReExport>,
-    pub(super) import_info: ModuleImports,
-    pub(super) docs: Option<String>,
+/// The three different kinds have slightly different rules on what syntax is allowed, as well as
+/// what operations can be performed in the body of procedures defined in the module. See the
+/// documentation for each variant for a summary of these differences.
+#[derive(Debug, Default, Copy, Clone, PartialEq, Eq)]
+#[repr(u8)]
+pub enum ModuleKind {
+    /// A library is a simple container of code that must be included into an executable module to
+    /// form a complete program.
+    ///
+    /// Library modules cannot use the `begin`..`end` syntax, which is used to define the
+    /// entrypoint procedure for an executable. Aside from this, they are free to use all other
+    /// MASM syntax.
+    #[default]
+    Library = 0,
+    /// An executable is the root module of a program, and provides the entrypoint for executing
+    /// that program.
+    ///
+    /// As the executable module is the root module, it may not export procedures for other modules
+    /// to depend on, it may only import and call externally-defined procedures, or private
+    /// locally-defined procedures.
+    ///
+    /// An executable module must contain a `begin`..`end` block.
+    Executable = 1,
+    /// A kernel is like a library module, but is special in a few ways:
+    ///
+    /// * Its code always executes in the root context, so it is stateful in a way that normal
+    /// libraries cannot replicate. This can be used to provide core services that would otherwise
+    /// not be possible to implement.
+    ///
+    /// * The procedures exported from the kernel may be the target of the `syscall` instruction,
+    /// and in fact _must_ be called that way.
+    ///
+    /// * Kernels may not use `syscall` or `call` instructions internally.
+    Kernel = 2,
```

Comment:
A thought for the future: I wonder if kernel modules should contain both exported procedures and executable code block (i.e., `begin ... end`). The exported procedures would work exactly as they do now. The executable code would be something that the VM starts executing in the root context.

The motivation for this is drawn from how we've implemented our transaction kernel [here](https://github.com/0xPolygonMiden/miden-base/tree/main/miden-lib/asm/kernels/transaction)). Basically, we define 2 MASM files: one for kernel "api" and another one for the kernel "main" function. It might be nice to combine these into a single unit. Though, this may require some pretty serious refactoring.

Reply:
I've had similar thoughts. In my case, it became quite clear (before even working on this PR), that what constitutes an "executable" is not a stable concept, as it depends heavily on the use case (not even the kernel necessarily, though I think the distinction is hard to draw, since the kernel is such a key part of a Miden program). As a result, executable modules aren't very useful as a distinct concept.

As further evidence of this, in the compiler (when targeting the rollup anyway), we are already planning to compile everything (note scripts, accounts) as library modules. This is despite the fact that note scripts are ostensibly "executable"-like.

I think it is useful to look at how existing assembly languages are used/integrate with the kernel/runtime/OS. In the ones I am familiar with, there is no distinction between different types of modules, only outputs, there is no namespacing (all symbols are defined in a global flat namespace, with the exception of explicitly-private symbols), and the assembler/linker is responsible for producing an appropriate artifact based on the desired target and flags. 

I think my recommendation of how we could modify Miden Assembly to address the weirdness around kernels/executables, while also providing some more useful tools for sharing assembled artifacts, can be summarized as follows:

1.) All modules are just containers for code (essentially the current notion of a "library" module).
2.) The notion of library namespace/path becomes syntactic sugar for defining symbols in a module such that they do not conflict in the global namespace. 
    * A module can be defined with no namespace/path (i.e. anonymously), in which case the procedures defined in that module are visible globally with those names.
    * Procedures defined in a "named" module shadow the global scope, so you can reference module-local procedures unqualified without ambiguity. 
    * We can introduce `::name` syntax to allow referencing an item in the global scope that has the same name as a local procedure. 
    * Module names and procedure names don't share the same semantic namespace, so it is fine to name a procedure the same thing as a module.
    * Similarly, imports become syntactic sugar for aliasing external procedure names to something more ergonomic locally. Without a corresponding import, procedure names are presumed to be fully-qualified. This doesn't change anything with regard to analysis, and is in fact easier for us to handle than the current import system.
3.) All procedures when compiled/assembled, are associated with a set of capabilities. By default this set is empty, but during compilation we detect needed capabilities based on the instructions used, and the capabilities of any procedures referenced. The initial capabilities I think we want to represent are: 
    * `read-only` (the procedure reads global memory, but does not write to it)
    * `write-only` (the procedure writes global memory, but does not read from it)
    * `read-write` (the procedure reads and writes global memory)
    * `syscall` (the procedure uses the `syscall` instruction)
    * `caller` (the procedure uses the `caller` instruction)
    * `kernel` (the procedure must be invoked via the `syscall` instruction)
4.) When assembling/compiling a set of modules, you must specify an artifact type: either `kernel`, `executable`, or `library`. These artifacts are associated with some metadata that describe their contents so that the assembler/VM can properly distinguish them, and allow the assembler to link against the artifacts it produces when compiling MASM code.
5.) A `kernel` artifact contains all of the code needed to implement the kernel, as well as the equivalent of crt0's `_start` function, i.e. the "true" entrypoint for a program. The kernel determines what happens from there, i.e. whether to invoke the `main` function of an executable, or something else, as defined by the runtime semantics implemented by the kernel. Only code which is not reachable from any of the exports, or the `_start` procedure, are candidates for dead-code elimination. All of the dependencies of the kernel must be provided when the kernel is assembled (i.e. it is not valid to reference a procedure that we do not have the implementation of on hand, however, we do not require that it be provided as MASM). The metadata associated with a `kernel` artifact is: the fact it is a kernel, the MAST root of the `_start` procedure, and the MAST roots of any syscalls provided by the kernel.
6.) An `executable` artifact must define a `main` symbol.  The `main` procedure is assumed to be the only entrypoint into the MAST of the executable, and as a result, we can aggressively dead-code eliminate anything not reachable from `main`. The metadata associated with the `executable` artifact is: the fact it is an executable, and the MAST root of the `main` procedure. How an executable is used depends on the kernel, so we may want to define executable targets more granularly (i.e. qualifying them with the kernel they expect to be run with)
7.) A `library` artifact has no restrictions on its definition (i.e. it can even define a `_start` function, or a `main` function, but it won't be treated specially). The only code that can be dead-code eliminated in a library is code that is not reachable from any of the exports, and is not itself exported. A library can be linked in to either a kernel or executable, with some conditions applied that I'll get to below. The metadata associated with a library is: the fact that it is a library, the MAST roots of all exported procedures of the library, the names of those procedures, and the procedure capabilities.
8.) A library artifact must be linked in to either a kernel or an executable to be useful. The capabilities of any procedures referenced from the library are validated to ensure they are valid for that artifact, i.e. if the kernel references a procedure `foo::bar`, then that procedure must not have the `syscall` capability, or an error will be raised).

So with these changes, a kernel gets a place where it can perform global initialization before any of its syscalls are invoked, and "owns" the executable lifecycle. We still distinguish between artifacts that are intended to be used like an executable, versus those that are intended to be used as libraries, which is useful from a tooling perspective. This also simplifies some aspects of compilation, and provides a path for lowering "traditional" languages to MASM directly, without sacrificing anything when hand-writing MASM.

I think this also ties in to the packaging/distribution issue. There are some things that I think we'd really like to be able to associate with the MAST we will produce from the assembler, that aren't there now. What I've outlined above I think are the most important parts, but we could maybe move some of it to the package (e.g. procedure names), but in that case, we would end up requiring the package in order to link against code in a library from a MASM module.

Anyway, if we're going to make some changes there, I think this is the direction we should go overall. The things I've outlined above are actually not particularly invasive changes (and in many cases may actually simplify things considerably), and can be picked off somewhat incrementally, it is largely a question of how exactly we want to draw boundaries between different artifact types, how they will be used with one another in practice (in terms of sharing artifacts, and linking against them when assembling a new artifact), and how we expect distribution to work (i.e. how do we plan to orchestrate running executables with a particular kernel, such that a kernel does not need to know about the specific executable).

> Though, this may require some pretty serious refactoring.

So removing the distinction between executable and kernel modules is actually pretty trivial, at least in the assembler. It barely exists as it is, so in many ways it simplifies some code. But as mentioned above, I think that there is still some value in being able to identify an executable vs a library.

The tiniest possible change would probably be to require/allow a kernel to define an entrypoint. A kernel which exposes an entrypoint could then be executed like an executable. When the VM is given an executable module, one of two things happens:

1.) If the kernel defines an entrypoint, then the executable module is loaded into the cache so that it is available, but it is the kernel entrypoint that gets invoked in the usual way. It is incumbent on the user to provide the inputs to the VM necessary to arrange for the kernel to execute the given executable module.
2.) If there is no kernel, or the kernel does not define an entrypoint, then the executable module is invoked in the usual way

This would probably make for another good issue to open for further discussion, though I do think these sorts of changes are pretty urgent since it will change the way users interact with MASM and the assembler (as will the packaging bits, so I kind of see those things as intertwined).